external-crates/move/crates/move-compiler/src/typing/translate.rs

```diff
@@ -2470,6 +2645,302 @@ fn make_arg_types<S: std::fmt::Display, F: Fn() -> S>(
+            m, &f, &param.value.name
+        )
+    };
+    subtype(context, call_loc, msg, tfun.clone(), param_ty);
+    // prefer the lambda type over the parameters to preserve annotations on the lambda
+    tfun
+}
+
+fn expand_macro(
+    context: &mut core::Context,
+    call_loc: Loc,
+    m: ModuleIdent,
+    f: FunctionName,
+    type_args: Vec<N::Type>,
+    args: Vec<macro_expand::Arg>,
+    return_ty: Type,
+) -> (Type, T::UnannotatedExp_) {
+    use T::SequenceItem_ as TS;
+    use T::UnannotatedExp_ as TE;
+
+    let valid = context.add_macro_expansion(m, f, call_loc);
+    if !valid {
+        assert!(context.env.has_errors());
+        return (context.error_type(call_loc), TE::UnresolvedError);
+    }
+    let res = match macro_expand::call(context, call_loc, m, f, type_args, args, return_ty) {
+        None => {
+            assert!(context.env.has_errors());
+            (context.error_type(call_loc), TE::UnresolvedError)
+        }
+        Some(macro_expand::ExpandedMacro {
+            by_value_args,
+            body,
+        }) => {
+            // bind the locals
+            let mut seq: VecDeque<_> = by_value_args
+                .into_iter()
+                .map(|(sp!(vloc, v_), e)| {
+                    let lvalue_ = match v_ {
+                        Some(var_) => N::LValue_::Var {
+                            mut_: None,
+                            var: sp(vloc, var_),
+                            unused_binding: false,
+                        },
+                        None => N::LValue_::Ignore,
+                    };
+                    let lvalue = sp(vloc, lvalue_);
+                    let lvalues = sp(vloc, vec![lvalue]);
+                    let b = bind_list(context, lvalues, Some(e.ty.clone()));
+                    let lvalue_ty = lvalues_expected_types(context, &b);
+                    sp(b.loc, TS::Bind(b, lvalue_ty, Box::new(e)))
+                })
+                .collect();
+            // add the body
+            let body = exp(context, body);
+            let ty = body.ty.clone();
+            seq.push_back(sp(body.exp.loc, TS::Seq(body)));
+            let use_funs = N::UseFuns::new(context.current_call_color());
+            let e_ = TE::Block((use_funs, seq));
+            (ty, e_)
+        }
+    };
+    context.pop_macro_expansion(&m, &f);
+    res
+}
+
+/// We need to make sure that arguments to macro calls are either lambdas or a Block
+/// These arguments are call-by-name so the whole expression is substituted in. So we need to track
+/// metadata about the scope where these expressions were originally written.
+/// The Block lets us track two pieces of metadata
+/// 1) We can track the use_fun_scope, which is used for resolving method calls correctly
+/// 2) After substitution, we can mark the Block as coming from a macro expansion which is used
+///    for tracking recursive macro calls
+fn block_macro_arg(context: &Context, sp!(loc, ne_): N::Exp) -> N::Exp {
```

Comment:
```suggestion
fn convert_macro_arg_to_block(context: &Context, sp!(loc, ne_): N::Exp) -> N::Exp {
```

Only because `block` can mean so many things.

Reply:
I like it. My first instinct was "blockify" :P (don't let me name things)